### PR TITLE
Leanback (Search and Recommendations) for AndroidTV

### PR DESCRIPTION
**Pull request by razzeee**
I finally had some time, to squash all the obvious bugs.
So here it is, it's basically picked from spmc and ported to v17 along with some improvements.
Big thanks to @koying 

The only drawback this has, is that it needs the webserver running on port 8080 (See: https://github.com/xbmc/xbmc/compare/master...Razzeee:leanback2?expand=1#diff-c9e9bcfe899ec6ab27a37e83a8b9a046R55 )
Otherwise you won't get any recommendations/voice search results.

### CODE REVIEW

**Comment by MartijnKaijser** [comment]
jenkins build this please

**Comment by Tolriq** [comment]
This require not only Webserver activated on port 8080 but no login and no password.

SPMC does allow port changing and authentication.

IMO asking users to activate Webserver without auth is a no go as real security risks. Specially with things like Shield that now have plex server on it and will tends users to open their box to internet.

File: tools/android/packaging/Makefile.in [hunk 1]
```diff
@@ -102,7 +102,7 @@ python: | xbmc/assets
 	cd xbmc/assets/python2.7/lib/python2.7/; rm -rf test config lib-dynload
 
 res:
-	mkdir -p xbmc/res xbmc/res/raw xbmc/res/values images
+	mkdir -p xbmc/gen/org/xbmc/kodi xbmc/res xbmc/res/raw xbmc/res/values images
```
Comment (wsnipex) [inline tools/android/packaging/Makefile.in]: xbmc/gen/org/xbmc/@APP_NAME_LC@

File: tools/android/packaging/Makefile.in [hunk 2]
```diff
@@ -112,9 +112,11 @@ res:
 	cp -fp media/drawable-xxhdpi/ic_launcher.png xbmc/res/drawable-xxhdpi/ic_launcher.png
 	cp -fp media/drawable-xhdpi/banner.png xbmc/res/drawable-xhdpi/banner.png
 	cp xbmc/strings.xml xbmc/res/values/
+	cp xbmc/colors.xml xbmc/res/values/
 	cp xbmc/activity_main.xml xbmc/res/layout/
+	cp xbmc/searchable.xml xbmc/res/xml/
 	mkdir -p tmp/res; $(AAPT) c -S xbmc/res -C tmp/res; cp -r -n xbmc/res tmp/ || true
-	$(AAPT) p -f -I $(SDKROOT)/platforms/$(SDK_PLATFORM)/android.jar -S tmp/res/ -M xbmc/AndroidManifest.xml -F images/@APP_NAME_LC@app-debug-skeleton.apk -J xbmc/src
+	$(AAPT) p -f -I $(SDKROOT)/platforms/$(SDK_PLATFORM)/android.jar -S tmp/res/ -M xbmc/AndroidManifest.xml -F images/@APP_NAME_LC@app-debug-skeleton.apk -J xbmc/gen/org/xbmc/kodi
```
Comment (wsnipex) [inline tools/android/packaging/Makefile.in]: @APP_NAME_LC@

File: tools/android/packaging/Makefile.in [hunk 3]
```diff
@@ -136,9 +139,9 @@ libs: $(PREFIX)/lib/@APP_NAME_LC@/lib@APP_NAME_LC@.so
 
 xbmc/classes.dex: res
 	mkdir -p xbmc/obj
-	@javac -classpath $(SDKROOT)/platforms/$(SDK_PLATFORM)/android.jar:xbmc/obj -d xbmc/obj -sourcepath xbmc/src xbmc/src/*.java
-	@javac -classpath $(SDKROOT)/platforms/$(SDK_PLATFORM)/android.jar:xbmc/obj -d xbmc/obj -sourcepath xbmc/src xbmc/src/org/xbmc/kodi/*.java
-	@$(DX) --dex --output=xbmc/classes.dex xbmc/obj
+	@javac -classpath $(SDKROOT)/platforms/$(SDK_PLATFORM)/android.jar:$(SDKROOT)/extras/android/support/v4/android-support-v4.jar:xbmc/obj -d xbmc/obj -sourcepath xbmc/src xbmc/gen/org/xbmc/kodi/*.java
+	@javac -classpath $(SDKROOT)/platforms/$(SDK_PLATFORM)/android.jar:$(SDKROOT)/extras/android/support/v4/android-support-v4.jar:xbmc/obj -d xbmc/obj -sourcepath xbmc/src xbmc/src/org/xbmc/kodi/*.java
```
Comment (wsnipex) [inline tools/android/packaging/Makefile.in]: @APP_NAME_LC@ in both lines

File: tools/android/packaging/xbmc/AndroidManifest.xml.in [hunk 4]
```diff
@@ -93,6 +93,25 @@
             </intent-filter>
         </receiver>
 
+        <provider
+            android:name=".XBMCImageContentProvider"
+            android:authorities="org.xbmc.kodi.image"
```
Comment (wsnipex) [inline tools/android/packaging/xbmc/AndroidManifest.xml.in]: @APP_NAME_LC@

File: tools/android/packaging/xbmc/AndroidManifest.xml.in [hunk 5]
```diff
@@ -93,6 +93,25 @@
...
+            android:exported="true" />
+
+        <provider
+            android:name=".XBMCMediaContentProvider"
+            android:authorities="org.xbmc.kodi.media"
```
Comment (wsnipex) [inline tools/android/packaging/xbmc/AndroidManifest.xml.in]: @APP_NAME_LC@

File: tools/android/packaging/xbmc/searchable.xml.in [hunk 6]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="utf-8"?>
+<searchable xmlns:android="http://schemas.android.com/apk/res/android"
+    android:label="@string/app_name"
+    android:hint="@string/search_hint" 
+
+    android:searchSuggestAuthority="org.xbmc.kodi.media"
+    android:searchSuggestPath="suggestions"
+    
+    android:searchSuggestIntentAction="android.intent.action.GET_CONTENT"
+    android:searchSuggestIntentData="content://org.xbmc.kodi.media/intent"
```
Comment (wsnipex) [inline tools/android/packaging/xbmc/searchable.xml.in]: not sure if this file is generated by the build sys, but @APP_NAME_LC@ applies here as well

File: tools/android/packaging/xbmc/src/org/xbmc/kodi/XBMCImageContentProvider.java.in [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+ * 
+ * See
+ * http://developer.android.com/reference/android/content/ContentProvider.html
+ * 
+ */
+public class XBMCImageContentProvider extends ContentProvider
+{
+  private static String TAG = "kodi";
+  
+  public static String AUTHORITY = "org.xbmc.kodi";
```
Comment (wsnipex) [inline tools/android/packaging/xbmc/src/org/xbmc/kodi/XBMCImageContentProvider.java.in]: @APP_NAME_LC@

File: tools/android/packaging/Makefile.in (hunk 1 above)
Comment (koying) [inline tools/android/packaging/Makefile.in]: Those are temp files. there is no added value to use APP_NAME, really.
To be paralleled with the src/org/xbmc/kodi/ folder, which is not parametrized (and does not need to)

**Comment by Razzeee** [comment]
Wasn't aware of the SPMC change, as I'm pretty out of the loop. Will pick those to this branch, when I find time.

**Comment by Razzeee** [comment]
Added two slightly edited commits from SPMC. Untested/builded so far.

File: tools/android/packaging/Makefile.in (hunk 1 above)
Comment (wsnipex) [inline tools/android/packaging/Makefile.in]: its doesn't hurt either. We use it everywhere, so it should be the same here, if only for the sake of consistency.

Comment (koying) [inline tools/android/packaging/Makefile.in]: Well, it hurts when it doesn't work. Took me hours to debug https://github.com/koying/SPMC/commit/eaedfe0470e46732cde14d93c72e97e9730260df , also for such parametrization which has not effect on the final product, but KODI_PREFIX was hardcoded somewhere.

So if we can avoid an unnecessary point of failure, let's do it ;)

Besides, org/xbmc/@APP_NAME_LC@ makes no sense, really, as a proper re-branding will have another domain name than xbmc.org. I added @APP_PACKAGE@ to properly handle that.

File: xbmc/video/windows/GUIWindowVideoNav.h [hunk 8]
```diff
@@ -49,7 +49,7 @@ class CGUIWindowVideoNav : public CGUIWindowVideoBase
   virtual bool OnAction(const CAction &action) override;
   virtual bool OnMessage(CGUIMessage& message) override;
 
-  virtual void OnItemInfo(const CFileItem& fileItem, ADDON::ScraperPtr &info) override;
+  virtual void OnItemInfo(CFileItem* fileItem, ADDON::ScraperPtr &info) override;
```
Comment (tamland) [inline xbmc/video/windows/GUIWindowVideoNav.h]: why is this changed to a raw pointer?

File: xbmc/utils/URIUtils.cpp [hunk 9]
```diff
@@ -211,6 +211,17 @@ void URIUtils::Split(const std::string& strFileNameAndPath,
...
+  // ignore options
+  i = strFileName.size() - 1;
+  while (i > 0)
+  {
+    char ch = strFileName[i];
+    if (ch == '?') break;
+    else i--;
+  }
+  if (i > 0)
+    strFileName = strFileName.substr(0, i);
```
Comment (tamland) [inline xbmc/utils/URIUtils.cpp]: `find_last_of` is what you were looking for:)

**Comment by MrMC** [comment]
If you are really going to require going through the internal webserver, you should pick up the black/white list clamps from MrMC fork. We block access to advancedsettings.xml, guisettings.xml, sources.xml and passwords.xml as these can contain clear text user/pass/ip information about other computers. Anyone that can hit the webserver can extract these file and exploit them.

tvOS platform does something as Android Leanback with TopShelf. TopShelf is an independent 'app' combined with the normal app. So one has to get clever on how to populate TopShelf with items and handle what happens when someone clicks on an item.

**Comment by Tolriq** [comment]
By default access to those files is locked down, blacklist could secure from users that create bad sources that open access to those files.

Using native code was rejected previously as no one would maintain it. What would be cool is to have a way like for nearly all the rest to allow Webserver to only run on localhost via a flag. Don't know if @koying have pushed on that direction yet.

**Comment by MrMC** [comment]
"By default access to those files is locked down", you should check again :) You can't get there direct but you can craft an url that will fetch them. Trust me on that.

Users are not prevented access via the file manager. Users with bad sources can fix them there.

As for native code, really ? What do you think all that java code is :)

What really is needed is a custom http socket that only runs on localhost, that handles json gets/puts. Oh wait, tcpserver does that already and you can now have multiple servers running. So crank up a dedicated tcpserver for leanback/topshelf type features.

**Comment by Tolriq** [comment]
Well then something went wrong at some point : https://github.com/xbmc/xbmc/blob/master/xbmc/utils/FileUtils.cpp#L110 was here to do that, maybe a call to that is missing somewhere, but no point in adding a new function that do the same thing IMO :)

About TCP, it does not allow image serving, so not an option actually for that need ;) By default there's already a localhost only JSONRPC server at 9090 ;)

About the full discussion on @Razzeee git was deleted I paste here the answer : 

My question : Why JSON and not native as you have the necessary skills for that ?

@koying answer :
1) That's what JSON is for
2) Overkill for a "nice-to-have" like recommendations, imo
3) That's additional code nobody but an Android/JNI specialist will be able to maintain
4) I'm lazy ;)

Anyway IMO JSON on HTTP for that is bad so you are preaching someone already convinced ;)

**Comment by MrMC** [comment]
I can pull anything under the app via tcp. Given the proper formatted url. I can even reach into the OS file space and pull/push anything that is not file permission protected. It's really an OMFG hole. Sure, there might be some ways around it but this closes a trivial exploit path.

zebra as it's both white and black :)

https://github.com/MrMC/mrmc/commit/955091c008e178e48aea737c891566c4e7e57359

That was the initial commit, it's been expanded since then.

**Comment by Tolriq** [comment]
Well I'm in holidays in a few hours :p But I'll see to add the missing calls to RemoteAccessAllowed where they should be. This will close the hole for short terms.

The problem of ZebraListAccessCheck is that its not precise enough on black list as it can block valid other files since not path check.

About TCP, you can't get file content from TCP server now, you can from Webserver it's another story. There's no url on TCP server just raw JSON.

**Comment by MartijnKaijser** [comment]
Can you keep this discussion elsewhere as this is drifting away from this PR. We all know know there are security gaps and the biggest one are users themselves as they install any crap they think they need. 
I want to get this PR in ASAP before window closes

**Comment by Tolriq** [comment]
The discussion is 100% related to this PR, Webserver is bad for this need as unsecure. 

Do not mix a known hole in webserver that the Team always says do not activate, with an option needed on Android TV that says : Hey activate the Webserver it allows cool things to happens like recommendation.

There's some dichotomy in the speech :)

So yes this discussion is needed and important as if known security problems are still present when you ask user to activate web server this is very very bad.

So either finish the speech about webserver security, or implement with something else in a secure way like native code, but right now there's some serious concerns for the end users.

**Comment by MrMC** [comment]
I'm done with talking here about websever/tcp security now :) This is the third time that I've brought it up only to get ponked down with, oh the users will not have free access to xyz. That's the point, there are some items/areas that users should ONLY get access via controlled methods.

A box inside a closed local net is one thing, but on a mobile phone/pad platform (android) that already has bad guys actively seeking pawnage ? I thought the motto was not to make their jobs easy.

**Comment by koying** [comment]
> A box inside a closed local net is one thing, but on a mobile phone/pad platform (android) that already has bad guys actively seeking pawnage ? I thought the motto was not to make their jobs easy.

Per definition, Android TV is a "local net" platform, though. 
But hey, just assume Kodi has zero security, and configure you lan accordingly. Exposing Kodi to the internet is, well, not a bright idea ;)

File: xbmc/video/windows/GUIWindowVideoNav.h (hunk 8 above)
Comment (Razzeee) [inline xbmc/video/windows/GUIWindowVideoNav.h]: Aligned it with how it is done for music and also it fixed a SIGSEGV that happend, not sure why exactly.

Comment (FernetMenta) [inline xbmc/video/windows/GUIWindowVideoNav.h]: not a good idea, IMO. I see no benefit in using this raw pointer here.

**Comment by Razzeee** [comment]
> If you are really going to require going through the internal webserver, you should pick up the black/white list clamps from MrMC fork. We block access to advancedsettings.xml, guisettings.xml, sources.xml and passwords.xml as these can contain clear text user/pass/ip information about other computers. Anyone that can hit the webserver can extract these file and exploit them.

Already on my todo list, along with various commits by koying or/and you. But I viewed it as a seperate PR, as this one might get stuck in review hell and the white/blacklist is something we want even if this PR is not wanted.

File: xbmc/video/windows/GUIWindowVideoNav.h (hunk 8 above)
Comment (koying) [inline xbmc/video/windows/GUIWindowVideoNav.h]: FWIW, it was a pointer before https://github.com/xbmc/xbmc/commit/f9186f1b511450afae47ba5f5ccfb862ebc7d834

Maybe @Razzeee stumbled on a bug with the am commit?

Comment (MrMC) [inline xbmc/video/windows/GUIWindowVideoNav.h]: watch using raw pointers to CFileItem's, most are really CFileItemPtr and are shared_ptr.  These have the tendency of slipping out from under you when you diddle with the raw pointer.

**Comment by Tolriq** [comment]
> Per definition, Android TV is a "local net" platform, though. 
But hey, just assume Kodi has zero security, and configure you lan accordingly. Exposing Kodi to the internet is, well, not a bright idea ;)

With Plex and such things less and less ;) You can't add a more than wanted feature rely on the user activating an option and thinking about the consequences that 99% are not even aware.

There's no big warning anywhere, there's no popup on webserver activation saying : Hey we open full access to your device, please think twice if you do not know what you are doing. Not even a message when password is empty (and it's empty by default ....)

Saying kodi have no security on github or hidden in the forum is one thing, telling your user when they open the door to everyone is another thing.

So please do not assume that what we know about internet and network and all the horrible hackers is known by someone that purchased a Gaming console like the shield and found Kodi cool for it's media purpose.

If the webserver road is chosen with known security problem this is Kodi becoming bad as pushing users to open unsecure door.

Known security problems must be fixed and / or very big warnings added, but done first.

**Comment by koying** [comment]
> With Plex and such things less and less

Well, having Plex working over the internet is not exactly piece of cake: https://support.plex.tv/hc/en-us/articles/200931138-Troubleshooting-Remote-Access
Unless you enable upnp port forwarding, which is a huge security hole by itself, and that is probably enabled by default in most of retail routers...

> Hey we open full access to your device

A tad exaggerated. You still have to forward 8080 from your router to the device, and it's not like you can do much harm with kodi webserver access, beside reading some config file that are or will be blacklisted...

File: xbmc/video/windows/GUIWindowVideoNav.h (hunk 8 above)
Comment (koying) [inline xbmc/video/windows/GUIWindowVideoNav.h]: Worked well enough < 17 ;)

Comment (Razzeee) [inline xbmc/video/windows/GUIWindowVideoNav.h]: I'll experiment with changing them back and also change the music part. If it works, no harm done, if not we need to trace the error.
See https://github.com/Razzeee/xbmc/pull/3#issuecomment-233133597 for a log excerpt

File: tools/android/packaging/Makefile.in (hunk 1 above)
Comment (Razzeee) [inline tools/android/packaging/Makefile.in]: You may also want to change "org" :feelsgood:

File: xbmc/video/windows/GUIWindowVideoNav.h (hunk 8 above)
Comment (FernetMenta) [inline xbmc/video/windows/GUIWindowVideoNav.h]: whatever it is going to be, it should not be hidden in a 1.5k loc change that suggests Android.

Comment (Razzeee) [inline xbmc/video/windows/GUIWindowVideoNav.h]: @FernetMenta 
your right obviously.

I reverted and can't seem to find the SIGSEGV. Might be solved due to other changes I did to fix the other bug I had at that stage. Thx to everybody pointing it out :)

File: xbmc/utils/URIUtils.cpp (hunk 9 above)
Comment (Razzeee) [inline xbmc/utils/URIUtils.cpp]: Changed, thx!

File: xbmc/video/dialogs/GUIDialogVideoInfo.h [hunk 10]
```diff
@@ -66,7 +66,7 @@ class CGUIDialogVideoInfo :
 
   static std::string GetLocalizedVideoType(const std::string &strType);
 
-  static void ShowFor(const CFileItem& item);
+  static void ShowFor(CFileItem item);
```
Comment (tamland) [inline xbmc/video/dialogs/GUIDialogVideoInfo.h]: missed one;)

File: xbmc/video/windows/GUIWindowVideoNav.cpp [hunk 11]
```diff
@@ -126,10 +126,78 @@ bool CGUIWindowVideoNav::OnMessage(CGUIMessage& message)
...
+              ADDON::ScraperPtr scrapper;
+              OnItemInfo(*pItem, scrapper);
+            }
+            break;
+          }
+        }
+        if (i >= m_vecItems->Size() && url.GetOption("showinfo") == "true")
+        {
+          std::string path = message.GetStringParam(0);
+          CFileItem* item = new CFileItem(path, URIUtils::HasSlashAtEnd(path));
```
Comment (tamland) [inline xbmc/video/windows/GUIWindowVideoNav.cpp]: Leaks. Just stack allocate it as there's no need for pointers here

File: xbmc/video/windows/GUIWindowVideoNav.cpp [hunk 12]
```diff
@@ -126,10 +126,78 @@ bool CGUIWindowVideoNav::OnMessage(CGUIMessage& message)
...
+              return tag;
+            };
+
+            *(item->GetVideoInfoTag()) = retrieveTag(m_vecItems->GetContent(), item->GetPath());
+            item->SetPath(item->GetVideoInfoTag()->m_strFileNameAndPath);
+          }
+          ADDON::ScraperPtr scrapper;
+          OnItemInfo(*item, scrapper);
+        }
+      }
```
Comment (tamland) [inline xbmc/video/windows/GUIWindowVideoNav.cpp]: Can you please refactor this? I have no idea what it does other than some parameter counting magic. (at least move it to some method out of sight.)

File: xbmc/video/dialogs/GUIDialogVideoInfo.h (hunk 10 above)
Comment (Razzeee) [inline xbmc/video/dialogs/GUIDialogVideoInfo.h]: Whoops, should be gone in a second.

File: xbmc/video/windows/GUIWindowVideoNav.cpp (hunk 11 above)
Comment (Razzeee) [inline xbmc/video/windows/GUIWindowVideoNav.cpp]: Can you get a little more into detail, my c++ is very weak and I don't know what you want me to do exactly. Thanks.

File: xbmc/video/windows/GUIWindowVideoNav.cpp (hunk 12 above)
Comment (Razzeee) [inline xbmc/video/windows/GUIWindowVideoNav.cpp]: Will do, this also needs a bunch of inline comments.

File: xbmc/video/windows/GUIWindowVideoNav.cpp (hunk 11 above)
Comment (tamland) [inline xbmc/video/windows/GUIWindowVideoNav.cpp]: it's the long way of saying change it to 
`CFileItem item(path, URIUtils::HasSlashAtEnd(path));` and adjust the `*`s below;)

File: tools/android/packaging/Makefile.in (hunk 1 above)
Comment (MartijnKaijser) [inline tools/android/packaging/Makefile.in]: i'd agree with koying. The @APP_NAME_LC@ needs to be revised in general as it doesn't replace the org.xbmc and we don't really want others using that either. So lets solve that in another PR

File: xbmc/video/windows/GUIWindowVideoNav.cpp [hunk 13]
```diff
@@ -130,67 +129,30 @@ bool CGUIWindowVideoNav::OnMessage(CGUIMessage& message)
       if (!CGUIWindowVideoBase::OnMessage(message))
         return false;
 
+      //Have we been called with a parameter attached?
```
Comment (Razzeee) [inline xbmc/video/windows/GUIWindowVideoNav.cpp]: @tamland please review this rewritten part, thanks

Comment (tamland) [inline xbmc/video/windows/GUIWindowVideoNav.cpp]: looks good

**Comment by Razzeee** [comment]
Rebased and squashed

**Comment by garbear** [comment]
Can we get a unit test for the URIUtils::Split() fix?

**Comment by Razzeee** [comment]
@garbear 
not sure what you mean, I included https://github.com/xbmc/xbmc/pull/10134/files#diff-c2c11e94617bec636884b0ccdf9dbaf6R127 already, that's pretty precisely what you asked for or am I getting you wrong?

**Comment by garbear** [comment]
yup, exactly. didn't see the unit test. good work

**Comment by openhardwarecoza** [comment]
Interesting round of public blackmouthing here https://plus.google.com/116630648530850689477/posts/MwNbePv7U8P amonst others.    Sad day for a great project.  Now someone mentions spmc,  let me go check that out.  Maybe their contributors doesn't hang around social media balckmouthing others (;

**Comment by Razzeee** [comment]
jenkins build this please

**Pull request closed without merging**